File: src/main/scala/com/pwootage/oc/js/duktape/DuktapeEngine.scala
package com.pwootage.oc.js.duktape

import com.pwootage.oc.js.JSEngine
import com.pwootage.oc.js.jsvalue.{JSStringValue, JSUndefined, JSValue}

class DuktapeEngine extends JSEngine {
  private var _maxMemory: Int = 0
  private var allocatedMemory: Int = 0
  private var _started = false

  override def setMaxMemory(max: Int): Unit = _maxMemory = max

  override def maxMemory: Int = _maxMemory

  override def usedMemory: Int = allocatedMemory

  override def started: Boolean = _started

  override def start(): Unit = {
    native_start()
  }

  override def destroy(): Unit = {
    native_destroy()
  }

  override def evalWithName(filename: String, js: String): JSValue = {
    //TODO
    println(s"NEED TO EXECUTE $filename")
    JSUndefined
  }

  override def executeThreaded(syncResult: JSValue): JSValue = {
    println(s"NEED TO EXECUTE THREADED")
    JSUndefined
  }

  //Native methods
  @native def native_start(): Unit

  @native def native_destroy(): Unit

}

File: src/main/java/com/pwootage/oc/js/duktape/DuktapeEngine.kt
package com.pwootage.oc.js.duktape

import com.pwootage.oc.js.*

class DuktapeEngine(val arch: DuktapeArchitecture) : JSEngine(arch) {
  private var duktapeEngineNative: Long = 0
  private var _started = false

  // Initialize on startup
  init {
    if (!DuktapeStatic.isInitialized) {
      DuktapeStatic.initialize()
    }
    native_start()
  }

  override fun setMaxMemory(max: Int) {
    native_set_max_memory(max)
  }

  override val maxMemory get() = native_get_max_memory()

  override val usedMemory get() = native_get_available_memory()

  override val started get() = _started

  override fun start() {
    _started = true
    val biosJS = StaticJSSrc.loadSrc("/assets/oc-js/bios/bios.js")
    val shim = StaticJSSrc.loadSrc("/assets/oc-js/es6-shim.min.js")
    val biosSrc =
      """$shim;(function(exports, global){try {$biosJS
           return "<bios ended execution>";
         /**/} catch (error) {
           const backup = new Error('<no error somehow>');
           return error.stack || backup.stack || '<no stack>';
         }
         })({}, this);
        """
    arch.componentInvoker.setResult(JSStringValue(biosSrc))
  }

  override fun destroy() {
    native_destroy()
  }

  override fun jsPushNext(jsNext: JSValue): JSValue = native_next(jsNext)

  //Native methods
  private external fun native_start(): Unit

  private external fun native_destroy(): Unit

  private external fun native_next(next: JSValue): JSValue

  private external fun native_set_max_memory(max: Int)

  private external fun native_get_max_memory(): Int

  private external fun native_get_available_memory(): Int
}